### PR TITLE
pin torch

### PR DESCRIPTION
**Pull request by yinweisu**
Checklist
* [ ] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [ ] Bumped the build number (if the version is unchanged)
* [ ] Reset the build number to `0` (if the version changed)
* [ ] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [ ] Ensured the license file is being packaged.

### CODE REVIEW

**Comment by yinweisu** [comment]
@conda-forge-admin, please rerender

**Comment by giswqs** [comment]
Here are the steps I took to debug the issue. 

First, create an `environment.yml` with the same dependency versions as the recipe without autogoluon.

```text
name: ag
channels:
    - conda-forge
dependencies:
    - python >=3.8,<3.11
    - numpy >=1.21,<1.27
    - scipy >=1.5.4,<1.12
    - pandas >=1.4.1,<1.6
    - scikit-learn >=1.0,<1.3
    - networkx >=3.0,<4
    - pytorch >=2.0,<2.1
    - lightgbm >=3.3,<3.4
    - xgboost >=1.6,<1.8
    - catboost >=1.1,<1.3
    - fastai >=2.3.1,<2.8
```

Then, create a conda env to test the dependencies. 

```bash
conda create -n ag python=3.10 -y
conda activate ag
conda install -c conda-forge mamba
mamba env update -n ag -f environment.yml
conda remove -n ag --all -y
```

Here is the output

image

That means the fastai package causes the issue. 
https://github.com/conda-forge/fastapi-feedstock/blob/main/recipe/meta.yaml

**Comment by giswqs** [comment]
It seems fastai supports 'pytorch>=1.7.0', so we should be able to update the fastai conda-forge recipe to remove the pytorch version upper bound <1.14

https://github.com/fastai/fastai/blob/master/environment.yml#L18C5-L18C17

**Comment by giswqs** [comment]
I have updated the fastai recipe here https://github.com/conda-forge/fastai-feedstock/pull/7. Once it is available on conda-forge, then we can rebuild the autogluon recipe.

**Comment by giswqs** [comment]
fastai 2.7.12 is available on conda-forge now. It might take 20-30 minutes for the CDN to propogate. 
https://anaconda.org/conda-forge/fastai/files

You can't simply comment out fastai in the recipe if the PyPI package requires it. The build will fail some some PyPI dependnecies are not listed in the recipe. 

Once is fastai 2.7.12 is available for install, you can revert the pytorch version pin to '>=1.7,<2.1' rather than '>=2.0,<2.1'. In that way, users can choose the pytorch version compatible with their env.

File: recipe/meta.yaml [hunk 1]
```diff
@@ -36,11 +36,11 @@ requirements:
     - pandas >=1.4.1,<1.6
     - scikit-learn >=1.0,<1.3
     - networkx >=3.0,<4
-    - pytorch >=1.13,<2.1
+    - pytorch >=2.0,<2.1
     - lightgbm >=3.3,<3.4
     - xgboost >=1.6,<1.8
     - catboost >=1.1,<1.3
-    - fastai >=2.3.1,<2.8
+    # - fastai >=2.3.1,<2.8
```
Comment (giswqs) [inline recipe/meta.yaml:43]: uncomment this line to add fastai back.

File: recipe/meta.yaml [hunk 2]
```diff
@@ -36,11 +36,11 @@ requirements:
     - pandas >=1.4.1,<1.6
     - scikit-learn >=1.0,<1.3
     - networkx >=3.0,<4
-    - pytorch >=1.13,<2.1
+    - pytorch >=2.0,<2.1
```
Comment (giswqs) [inline recipe/meta.yaml:39]: change it back to '>=1.13' so that autogluon can be installed in env with pytorch <2.0 and newer versions.

**Comment by giswqs** [comment]
CI has passed. Merging now. It should now support pytorch 2.0

**Pull request merged**